java/common/Reference.java

```diff
@@ -204,21 +210,33 @@ public boolean equals(Object o) {
     public static class Label extends Reference {
 
         private final String label;
+        private final String scope;
+
         private final int hash;
 
-        Label(String label) {
+        Label(String label, @Nullable String scope) {
             super(Type.LABEL, false);
             this.label = label;
-            this.hash = Objects.hash(this.type, this.isVisible, this.label);
+            this.scope = scope;
+            this.hash = Objects.hash(this.type, this.isVisible, this.label, this.scope);
         }
 
         public String label() {
             return label;
         }
 
+        public Optional<String> scope() {
+            return Optional.ofNullable(scope);
+        }
+
+        public String scopedLabel() {
+            if (scope == null) return label;
+            else return scope + COLON + label;
+        }
+
         @Override
         public String name() {
-            return TypeQLToken.Char.UNDERSCORE.toString() + label;
+            return scopedLabel();
```

Comment:
The name of a labeled variable is also now just a label rather than being preceded by a `_`